fuel-types/src/canonical.rs

```diff
@@ -0,0 +1,619 @@
+
+        Ok(())
+    }
+}
+
+impl<const N: usize, T: Deserialize> Deserialize for [T; N] {
+    fn decode_static<I: Input + ?Sized>(buffer: &mut I) -> Result<Self, Error> {
+        if Self::UNALIGNED_BYTES {
+            let mut bytes: [u8; N] = [0; N];
+            buffer.read(bytes.as_mut())?;
+            buffer.skip(alignment_bytes(N))?;
+            let ref_typed: &[T; N] = unsafe { core::mem::transmute(&bytes) };
+            let typed: [T; N] = unsafe { core::ptr::read(ref_typed) };
+            Ok(typed)
+        } else {
+            // Spec doesn't say how to deserialize arrays with unaligned
+            // primitives(as `u16`, `u32`, `usize`), so unpad them.
+            let mut uninit = <MaybeUninit<[T; N]>>::uninit();
+            // The following line coerces the pointer to the array to a pointer
+            // to the first array element which is equivalent.
+            let mut ptr = uninit.as_mut_ptr() as *mut T;
+            for _ in 0..N {
+                let decoded = T::decode_static(buffer)?;
+                // SAFETY: We do not read uninitialized array contents
+                // 		 while initializing them.
+                unsafe {
+                    core::ptr::write(ptr, decoded);
+                }
+                // SAFETY: Point to the next element after every iteration.
+                // 		 We do this N times therefore this is safe.
+                ptr = unsafe { ptr.add(1) };
+            }
+            // SAFETY: All array elements have been initialized above.
+            let init = unsafe { uninit.assume_init() };
+            Ok(init)
+        }
+    }
+
+    fn decode_dynamic<I: Input + ?Sized>(&mut self, buffer: &mut I) -> Result<(), Error> {
+        for e in self.iter_mut() {
+            e.decode_dynamic(buffer)?;
+        }
+        Ok(())
+    }
+}
+
+impl Output for Vec<u8> {
+    fn write(&mut self, bytes: &[u8]) -> Result<(), Error> {
+        self.extend_from_slice(bytes);
+        Ok(())
+    }
+}
+
+impl<'a> Output for &'a mut [u8] {
+    fn write(&mut self, from: &[u8]) -> Result<(), Error> {
+        if from.len() > self.len() {
+            return Err(Error::BufferIsTooShort)
+        }
+        let len = from.len();
+        self[..len].copy_from_slice(from);
+        // We need to reduce the inner slice by `len`, because we already filled them.
+        let reduced = &mut self[len..];
+
+        // Compiler is not clever enough to allow it.
+        // https://stackoverflow.com/questions/25730586/how-can-i-create-my-own-data-structure-with-an-iterator-that-returns-mutable-ref
+        *self = unsafe { &mut *(reduced as *mut [u8]) };
+        Ok(())
+    }
+}
+
+/// Counts the number of written bytes.
+pub struct SizeCalculator(usize);
+
+impl SizeCalculator {
+    /// The number of written bytes.
+    pub fn size(self) -> usize {
+        self.0
+    }
+}
+
+impl Output for SizeCalculator {
+    fn write(&mut self, bytes: &[u8]) -> Result<(), Error> {
+        self.0 = self
```

Comment:
How does this compare to the previous approach of using the memlayout + some manual logic for dynamic sizes? Are we incurring any new overhead by first traversing all serializable fields, converting each one into a slice, and then storing each of their lengths here?

Reply:
If the process of converting a field into a `&[u8]` has overhead, then we end up paying this twice for every call to `to_bytes()` since we get the size first to initialize a vec with enough capacity.

Reply:
Maybe we should consider calculating the static size of each type at compile time?

Reply:
It is the safest way to know the final size of the object. But yeah, we can optimize it later by introducing `STATIC_PART_SIZE`.

I'm okay with this during the first iteration of this transition.

Reply:
Static size *is* computed into `SIZE_STATIC` constant. The transition away from `SizeCalculator´ shoulnd't be too hard. I left that for a follow-up PR, but it might make sense to do it right now.

Reply:
It did make sense to do right now, as it required so much changes.